eth/beacon/state_machines/forks/serenity/state_transitions.py

```diff
@@ -16,21 +20,29 @@ def __init__(self, config: BeaconConfig):
         self.config = config
 
     def apply_state_transition(self, state: BeaconState, block: BaseBeaconBlock) -> BeaconState:
-        state = self.per_slot_transition(state, block)
-        state = self.per_block_transition(state, block)
-        state = self.per_epoch_transition(state, block)
+        while state.slot != block.slot:
```

Comment:
Could do something like the following:

```python
for _ in range(state.slot, block.slot):
    do_loop()

if state.slot != block.slot:
    raise SomeException("`state.slot` must be equal to `block.slot` after a state transition")

Reply:
What type of exception would you raise here?
This would be an error in the software, not in values provided by the user.

Reply:
We typically use 

```
raise Exception("Invariant: It should be impossible for the `state.slot` to not be equal to the `block.slot` after running the transitions logic stuff....")
```

Just something to signify that 1) when reading that the code path *should* be unreachable and 2) if anyone ever encounters the exception that there is very likely a logic flaw in the code which needs to be addressed.

Reply:
How about:

```
for _ in range(state.slot, block.slot):
    ... # do loop
    if state.slot == block.slot:
        break
else:
    raise Exception("Invariant: ...")
```

Functionally equivalent but I think it might be a bit clearer what is happening.

Reply:
huh. Didn't know `for` loops had an optional `else`.
That looks good to me. Will make an issue

Reply:
Little known feature.  It's only triggered if the loop exits naturally.  If it exits due to a `break` call it isn't executed.  You can do the same with `while` loops.